django/utils/dateformat.py

```diff
@@ -278,7 +286,15 @@ def o(self):
 
     def r(self):
         "RFC 5322 formatted date; e.g. 'Thu, 21 Dec 2000 16:01:07 +0200'"
-        return self.format('D, j M Y H:i:s O')
+        if not self.timezone:
+            raise TypeError("The 'r' format specifier can only be used on datetime objects")
```

Comment:
Do we need this? I don't see any failures without it, I think we can simplify this method, e.g.
```python
def r(self):
    if self.timezone and is_naive(self.data):
         self.data = make_aware(self.data, timezone=self.timezone)
    return format_datetime_rfc5322(self.data)
```

Reply:
The `raise TypeError(...)` was leftover from a previous approach I'd tried, I've removed it.

Your suggested approach of mutating `self.data` could lead to problems because the `self.data` object is reused for all the format characters passed to `.format()`.

Consider the following example: `format(datetime.now(), 'U r U')`.
The first `U` format will receive a naive datetime object, then the `r` format will mutate that value to be an aware one, then finally the second `U` format will receive the new aware object.

I'm not sure if it can cause actual problems in practice but it's giving me a bad feeling (and none of the other format methods manipulate `self.data`).

Reply:
Agreed :+1: